sacc2/covariance.py

```diff
@@ -0,0 +1,484 @@
+    """A covariance subclass representing block diagonal covariances
+
+    Block diagonal covariances have sub-blocks that are full dense matrices,
+    but without correlations between the blocks. This feature can be taken
+    advantage of when doing matrix operations like multiplication or inversion.
+
+    Attributes
+    ----------
+    blocks: list[arrays]
+        list of sub-blocks of the matrix
+    block_sizes: list[int]
+        list of sizes n of each the n x n sub-blocks
+    size: int
+        overall total size of the matrix
+
+    """
+    def __init__(self, blocks):
+        """Create a BlockDiagonalCovariance object from a list of blocks
+
+        Parameters
+        ----------
+        blocks: sequence of arrays
+            List or other sequence of the sub-matrices
+
+        """
+        self.blocks = [np.atleast_2d(B) for B in blocks]
+        self.block_sizes = [len(B) for B in self.blocks]
+        self.size = sum(self.block_sizes)
+
+    def to_hdu(self):
+        """Write a FITS HDU from the data, ready to be saved.
+
+        The data in the HDU is stored as a single 1 x size image,
+        and the header contains the information needed to reconstruct it.
+
+        Returns
+        -------
+
+        hdu: astropy.fits.ImageHDU object
+            HDU containing data and metadata
+
+        """
+        hdu = fits.ImageHDU(np.concatenate([b.flatten() for b in self.blocks]))
+        hdu.name = 'covariance'
+        hdu.header['sacctype'] = 'cov'
+        hdu.header['saccclss'] = self.cov_type
+        hdu.header['size'] = self.size
+        hdu.header['blocks'] = len(self.blocks)
+        for i, s in enumerate(self.block_sizes):
+            hdu.header[f'size_{i}'] = s
+        return hdu
+
+    @classmethod
+    def from_hdu(cls, hdu):
+        """Read a covariance object from a loaded FITS HDU.
+
+        Parameters
+        ----------
+        hdu: FITS HDU object as read in by astropy.
+
+        Returns
+        -------
+
+        cov: BlockDiagonalCovariance
+            Loaded covariance object
+
+        """
+        n = hdu.header['blocks']
+        block_sizes = [hdu.header[f'size_{i}'] for i in range(n)]
+        s = 0
+
+        blocks = []
+        for b in block_sizes:
+            B = hdu.data[s:s + b**2].reshape((b, b))
+            s += b**2
+            blocks.append(B)
+        return cls(blocks)
+
+    def get_block(self, indices):
+        """Read a (not necessarily contiguous) sublock of the matrix
+
+        Parameters
+        ----------
+
+        indices: array
+            An array of integer indices, which must be in 
+            ascending order
+        
+        Returns
+        -------
+
+        cov: array
+            A full (dense) 2x2 array of the submatrix.
+
+        """
+        if np.any(np.diff(indices))<0:
+            raise ValueError("Indices passed to BlockDiagonalCovariance.get_block must be in ascending order")
```

Comment:
No, it's unsafe, because you would normally use this to select a chunk of the mean as well as the matrix, so if you just reordered that matrix it would return inconsistent data.

So the caller needs to sort them (and in the places we call it already this is automatic).

Reply:
It could be rewritten to allow arbitrary orderings, but the code and the resulting matrix would be very messy.  It seemed better to disallow this.